style/layer/road.js

```diff
@@ -0,0 +1,367 @@
+    );
+  }
+}
+
+class TrunkTunnel extends Trunk {
+  constructor() {
+    super();
+    this.brunnel = "tunnel";
+    this.fillColor = `hsl(${this.hue}, 41%, 90%)`;
+  }
+}
+
+class MotorwayLinkTunnel extends MotorwayLink {
+  constructor() {
+    super();
+    this.brunnel = "tunnel";
+    this.casingColor = this.colorCasingLowZoom.concat(
+      minzoomBrunnel + 0.5,
+      `hsl(${this.hue}, 71%, 75%)`
+    );
+    this.fillColor = this.colorFillLowZoom.concat(
+      minzoomBrunnel + 0.5,
+      `hsl(${this.hue}, 71%, 90%)`
+    );
+  }
+}
+
+class TrunkLinkTunnel extends TrunkLink {
+  constructor() {
+    super();
+    this.brunnel = "tunnel";
+    this.fillColor = `hsl(${this.hue}, 41%, 90%)`;
+  }
+}
+
+/*
+ Road style generation helper functions
+*/
+
+function roadPaint(color, width) {
+  return {
+    "line-color": color,
+    "line-width": {
+      base: roadExp,
+      stops: width,
+    },
+    "line-blur": 0.5,
+  };
+}
+
+function tunnelCasePaint(color, width) {
+  return {
+    "line-color": color,
+    "line-width": {
+      base: roadExp,
+      stops: width,
+    },
+    "line-opacity": 1,
+    "line-dasharray": tunDashArray,
+  };
+}
+
+//Helper function to create a "filter" block for a particular road class.
+function filterRoad(roadClass, ramp, brunnel) {
+  return [
+    "all",
+    brunnel === "surface"
+      ? ["!in", "brunnel", "bridge", "tunnel"]
+      : ["==", "brunnel", brunnel],
+    ["==", "class", roadClass],
+    [ramp ? "==" : "!=", "ramp", 1],
+  ];
+}
+
+//Base definition that applies to all roads (fill and casing).
+var defRoad = {
+  type: "line",
+  source: "openmaptiles",
+  "source-layer": "transportation",
+};
+
+//Generate a unique layer ID
+function uniqueLayerID(hwyClass, link, part, brunnel) {
+  return [hwyClass, link ? "link" : "road", part, brunnel].join("_");
+}
+
+function baseRoadLayer(highwayClass, id, brunnel, minzoom, link) {
+  var layer = layerClone(
+    defRoad,
+    uniqueLayerID(highwayClass, link, id, brunnel)
+  );
+  layer.filter = filterRoad(highwayClass, link, brunnel);
+  layer.minzoom = minzoom;
+  return layer;
+}
+
+const Road = {
+  fill: function () {
+    var layer = baseRoadLayer(
+      this.highwayClass,
+      "fill",
+      this.brunnel,
+      this.minZoomFill,
+      this.link
+    );
+    layer.layout = layoutRoadFill;
+    layer.paint = roadPaint(this.fillColor, this.fillWidth);
+    return layer;
+  },
+  casing: function () {
+    var layer = baseRoadLayer(
+      this.highwayClass,
+      "casing",
+      this.brunnel,
+      this.minZoomCasing,
+      this.link
+    );
+    layer.layout = layoutRoadCase;
+    if (this.brunnel === "tunnel") {
+      layer.paint = tunnelCasePaint(this.casingColor, this.casingWidth);
+    } else {
+      layer.paint = roadPaint(this.casingColor, this.casingWidth);
+    }
+    return layer;
+  },
+};
+
+Object.setPrototypeOf(Motorway.prototype, Road);
+Object.setPrototypeOf(Trunk.prototype, Road);
```

Comment:
Road is a constant, not a class so I can't just extend it, at least the way it's currently implemented.

Reply:
Changing Road to a class ought to work just fine.

(There might be more complex cases that call for using mixins such as this, but it would be difficult to analyze statically.)

Reply:
Cool, that totally worked!